packages/headless/src/app/state-key.ts

```diff
@@ -1 +1,24 @@
+import type {CoreEngineNext} from './engine';
+
 export const stateKey = Symbol('state');
+
+export const redactEngine = <TEngine extends CoreEngineNext>(
```

Comment:
This method could (and probably will) also directly include the prototype chain shenanigans.

After that, when the prototype chain guards are in place on the controllers, we could remove/reformulate the warning to be more of a "How the heck did you reach this? Please contact Coveo Support."